File: app/src/main/java/com/meli/android/carddrawer/app/MainActivity.java
package com.meli.android.carddrawer.app;

import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.util.Log;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Spinner;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.SwitchCompat;
import com.meli.android.carddrawer.app.model.CardComposite;
import com.meli.android.carddrawer.app.model.CustomAccountMoneyConfiguration;
import com.meli.android.carddrawer.app.model.HybridCreditConfiguration;
import com.meli.android.carddrawer.app.model.MasterCardConfiguration;
import com.meli.android.carddrawer.app.model.PixConfiguration;
import com.meli.android.carddrawer.app.model.UrlTestConfiguration;
import com.meli.android.carddrawer.app.model.VisaCardBlueConfiguration;
import com.meli.android.carddrawer.app.model.VisaCardGrayConfiguration;
import com.meli.android.carddrawer.app.model.VisaCardGreenConfiguration;
import com.meli.android.carddrawer.app.model.VisaCardRedConfiguration;
import com.meli.android.carddrawer.app.model.VisaCardYellowConfiguration;
import com.meli.android.carddrawer.configuration.CardDrawerStyle;
import com.meli.android.carddrawer.configuration.DefaultCardConfiguration;
import com.meli.android.carddrawer.model.CardDrawerSource;
import com.meli.android.carddrawer.model.CardDrawerView;
import com.meli.android.carddrawer.model.customview.CardDrawerSwitch;
import com.meli.android.carddrawer.model.customview.SwitchModel;
import java.util.ArrayList;
import java.util.List;

public class MainActivity extends AppCompatActivity {

    /* default */ CardComposite card;
    /* default */ CardDrawerView cardDrawerView;
    /* default */ CardDrawerView cardDrawerViewLowRes;
    /* default */ CardDrawerView cardDrawerViewMedium;
    /* default */ CardDrawerView cardDrawerViewMediumRes;
    /* default */ SwitchCompat switchCustomView;

    @Override
    protected void onCreate(final Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.card_drawer_app_activity_main);

        cardDrawerView = findViewById(R.id.card_header_container);
        cardDrawerViewLowRes = findViewById(R.id.card_header_lowres_container);
        cardDrawerViewMedium = findViewById(R.id.card_header_medium_container);
        cardDrawerViewMediumRes = findViewById(R.id.card_header_mediumres_container);
        switchCustomView = findViewById(R.id.card_drawer_custom_view_switch);

        card = new CardComposite();
        card.addCard(cardDrawerView.getCard());
        card.addCard(cardDrawerViewLowRes.getCard());
        card.addCard(cardDrawerViewMedium.getCard());
        card.addCard(cardDrawerViewMediumRes.getCard());
        ((SwitchCompat) findViewById(R.id.card_header_switch_responsive)).setOnCheckedChangeListener(
            (buttonView, isChecked) -> {
                final int behaviour = isChecked ?
                    CardDrawerView.Behaviour.RESPONSIVE : CardDrawerView.Behaviour.REGULAR;
                cardDrawerView.setBehaviour(behaviour);
                cardDrawerViewLowRes.setBehaviour(behaviour);
                cardDrawerViewMedium.setBehaviour(behaviour);
                cardDrawerViewMediumRes.setBehaviour(behaviour);
            });

        final SwitchCompat switchLowres = findViewById(R.id.card_header_lowres_switch);
        final SwitchCompat switchMedium = findViewById(R.id.card_header_medium_switch);
        final SwitchCompat switchMediumRes = findViewById(R.id.card_header_mediumres_switch);

        switchLowres.setOnCheckedChangeListener(
            (buttonView, isChecked) -> {
                if (isChecked) {
                    switchMedium.setChecked(false);
                    switchMediumRes.setChecked(false);
                    cardDrawerViewLowRes.setVisibility(View.VISIBLE);
                    cardDrawerViewMedium.setVisibility(View.GONE);
                    cardDrawerViewMediumRes.setVisibility(View.GONE);
                    cardDrawerView.setVisibility(View.GONE);
                } else {
                    cardDrawerViewLowRes.setVisibility(View.GONE);
                    if (switchMedium.isChecked()) {
                        cardDrawerViewMedium.setVisibility(View.VISIBLE);
                    } else {
                        if (switchMediumRes.isChecked()) {
                            cardDrawerViewMediumRes.setVisibility(View.VISIBLE);
                        } else {
                            cardDrawerView.setVisibility(View.VISIBLE);
                        }
                    }
                }
            });

        switchMedium.setOnCheckedChangeListener(
            (buttonView, isChecked) -> {
                if (isChecked) {
                    switchLowres.setChecked(false);
                    switchMediumRes.setChecked(false);
                    cardDrawerViewMedium.setVisibility(View.VISIBLE);
                    cardDrawerView.setVisibility(View.GONE);
                    cardDrawerViewLowRes.setVisibility(View.GONE);
                    cardDrawerViewMediumRes.setVisibility(View.GONE);
                } else {
                    cardDrawerViewMedium.setVisibility(View.GONE);
                    if (switchLowres.isChecked()) {
                        cardDrawerViewLowRes.setVisibility(View.VISIBLE);
                    } else {
                        if (switchMediumRes.isChecked()) {
                            cardDrawerViewMediumRes.setVisibility(View.VISIBLE);
                        } else {
                            cardDrawerView.setVisibility(View.VISIBLE);
                        }
                    }
                }
            });

        switchMediumRes.setOnCheckedChangeListener(
                (buttonView, isChecked) -> {
                    if (isChecked) {
                        switchLowres.setChecked(false);
                        switchMedium.setChecked(false);
                        cardDrawerViewMediumRes.setVisibility(View.VISIBLE);
                        cardDrawerView.setVisibility(View.GONE);
                        cardDrawerViewMedium.setVisibility(View.GONE);
                        cardDrawerViewLowRes.setVisibility(View.GONE);
                    } else {
                        cardDrawerViewMediumRes.setVisibility(View.GONE);
                        if (switchLowres.isChecked()) {
                            cardDrawerViewLowRes.setVisibility(View.VISIBLE);
                        } else {
                            if (switchMedium.isChecked()) {
                                cardDrawerViewMedium.setVisibility(View.VISIBLE);
                            } else {
                                cardDrawerView.setVisibility(View.VISIBLE);
                            }
                        }
                    }
                });

        final SwitchCompat switchDisabled = findViewById(R.id.card_header_disabled_switch);
        switchDisabled.setOnCheckedChangeListener((buttonView, isChecked) -> {
            cardDrawerView.setEnabled(!isChecked);
            cardDrawerViewLowRes.setEnabled(!isChecked);
            cardDrawerViewMedium.setEnabled(!isChecked);
            cardDrawerViewMediumRes.setEnabled(!isChecked);
        });

        initCardConfigurationOptions();
        initCardNumber();
        initCardName();
        initExpirationDate();
        initSecurityCode();
    }

    private void initSecurityCode() {
        final TextView securityCode = findViewById(R.id.security_code);

        securityCode.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(final CharSequence charSequence, final int i, final int i1, final int i2) {
                cardDrawerView.showSecurityCode();
                cardDrawerViewLowRes.showSecurityCode();
                cardDrawerViewMedium.showSecurityCode();
                cardDrawerViewMediumRes.showSecurityCode();
            }

            @Override
            public void onTextChanged(final CharSequence charSequence, final int i, final int i1, final int i2) {
                card.setSecCode(charSequence.toString());
            }

            @Override
            public void afterTextChanged(final Editable editable) {
                if (editable.toString().isEmpty()) {
                    cardDrawerView.show();
                    cardDrawerViewLowRes.show();
                    cardDrawerViewMedium.show();
                    cardDrawerViewMediumRes.show();
                }
            }
        });
    }

    private void initExpirationDate() {
        final TextView expirationDate = findViewById(R.id.expiration_date);

        expirationDate.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(final CharSequence charSequence, final int i, final int i1, final int i2) {
                //Do nothing
            }

            @Override
            public void onTextChanged(final CharSequence charSequence, final int i, final int i1, final int i2) {
                card.setExpiration(charSequence.toString());
            }

            @Override
            public void afterTextChanged(final Editable editable) {
                //Do nothing
            }
        });
    }

    private void initCardName() {
        final TextView cardName = findViewById(R.id.card_name);

        cardName.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(final CharSequence charSequence, final int i, final int i1, final int i2) {
                //Do nothing
            }

            @Override
            public void onTextChanged(final CharSequence charSequence, final int i, final int i1, final int i2) {
                card.setName(charSequence.toString());
            }

            @Override
            public void afterTextChanged(final Editable editable) {
                //Do nothing
            }
        });

        cardName.setOnClickListener(v -> {
            cardDrawerView.show();
            cardDrawerViewLowRes.show();
            cardDrawerViewMedium.show();
            cardDrawerViewMediumRes.show();
        });
    }

    private void initCardNumber() {
        final TextView cardNumber = findViewById(R.id.card_number);

        cardNumber.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(final CharSequence charSequence, final int i, final int i1, final int i2) {
                //Do nothing
            }

            @Override
            public void onTextChanged(final CharSequence charSequence, final int i, final int i1, final int i2) {
                card.setNumber(charSequence.toString());
            }

            @Override
            public void afterTextChanged(final Editable editable) {
                //Do nothing
            }
        });

        cardNumber.setOnClickListener(v -> {
            cardDrawerView.show();
            cardDrawerViewLowRes.show();
            cardDrawerViewMedium.show();
            cardDrawerViewMediumRes.show();
        });
    }

    private void initCardConfigurationOptions() {
        final Spinner cardsSpinner = findViewById(R.id.spinner_cards);

        final List<CardConfigurationOption> cardOptions = new ArrayList<>();

        cardOptions.add(new CardConfigurationOption("Default", new DefaultCardConfiguration(this)));
        cardOptions.add(new CardConfigurationOption("Visa blue", new VisaCardBlueConfiguration(this)));
        cardOptions.add(new CardConfigurationOption("Visa green", new VisaCardGreenConfiguration(this)));
        cardOptions.add(new CardConfigurationOption("Visa red", new VisaCardRedConfiguration(this)));
        cardOptions.add(new CardConfigurationOption("Visa gray", new VisaCardGrayConfiguration(this)));
        cardOptions.add(new CardConfigurationOption("Visa yellow", new VisaCardYellowConfiguration(this)));
        cardOptions.add(new CardConfigurationOption("Master", new MasterCardConfiguration(this)));
        cardOptions.add(new CardConfigurationOption("Url Test", new UrlTestConfiguration(this)));
        cardOptions.add(new CardConfigurationOption("Hybrid Account Money", CardDrawerStyle.ACCOUNT_MONEY_HYBRID));
        cardOptions.add(new CardConfigurationOption("Hybrid Credit", new HybridCreditConfiguration()));
        cardOptions.add(new CardConfigurationOption("Default Account Money", CardDrawerStyle.ACCOUNT_MONEY_DEFAULT));
        cardOptions.add(new CardConfigurationOption("Custom account money", new CustomAccountMoneyConfiguration()));
        cardOptions.add(new CardConfigurationOption("PIX", new PixConfiguration(this)));

        final ArrayAdapter<CardConfigurationOption> cardAdapter =
            new ArrayAdapter<>(this, android.R.layout.simple_spinner_item, cardOptions);
        cardAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        cardsSpinner.setAdapter(cardAdapter);
        cardsSpinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(final AdapterView<?> parent, final View view, final int position, final long id) {
                final CardConfigurationOption selection = cardOptions.get(position);
                final CardDrawerSource source = selection.getCardConfiguration();
                final CardDrawerStyle style = selection.getCardStyle();
                if (source != null) {
                    cardDrawerView.show(source);
                    cardDrawerViewLowRes.show(source);
                    cardDrawerViewMedium.show(source);
                    cardDrawerViewMediumRes.show(source);
                } else if (style != null) {
                    cardDrawerView.setStyle(style);
                    cardDrawerViewLowRes.setStyle(style);
                    cardDrawerViewMedium.setStyle(style);
                    cardDrawerViewMediumRes.setStyle(style);
                }

                setUpCustomView(style, switchCustomView.isChecked());
                switchCustomView.setOnCheckedChangeListener((buttonView, isChecked) -> setUpCustomView(style, isChecked));
            }

            private void setUpCustomView(@NonNull final CardDrawerStyle style, @NonNull final Boolean isChecked) {
                if (isChecked) {
                    final SwitchModel model;
                    if (style == CardDrawerStyle.ACCOUNT_MONEY_HYBRID) {
                        model = SwitchFactoryModelSample.createModelForAccountMoneyHybrid();
                    } else {
                        model = SwitchFactoryModelSample.createModel();
                    }
                    final CardDrawerSwitch viewHighRes = new CardDrawerSwitch(MainActivity.this);
                    viewHighRes.setSwitchListener(optionId -> Log.i("CARD_DRAWER", "ID: "+ optionId));
                    viewHighRes.setSwitchModel(model);
                    cardDrawerView.setCustomView(viewHighRes);
                    final CardDrawerSwitch viewLowRes = new CardDrawerSwitch(MainActivity.this);
                    viewLowRes.setSwitchListener(optionId -> Log.i("CARD_DRAWER", "ID: "+ optionId));
                    viewLowRes.setSwitchModel(model);
                    viewLowRes.setConfiguration(cardDrawerViewLowRes.buildCustomViewConfiguration());
                    cardDrawerViewLowRes.setCustomView(viewLowRes);
                    final CardDrawerSwitch viewMediumRes = new CardDrawerSwitch(MainActivity.this);
                    viewMediumRes.setSwitchListener(optionId -> Log.i("CARD_DRAWER", "ID: "+ optionId));
                    viewMediumRes.setSwitchModel(model);
                    viewMediumRes.setConfiguration(cardDrawerViewMediumRes.buildCustomViewConfiguration());
                    cardDrawerViewMediumRes.setCustomView(viewMediumRes);
                } else {
                    cardDrawerView.setCustomView(null);
                    cardDrawerViewLowRes.setCustomView(null);
                    cardDrawerViewMediumRes.setCustomView(null);
                }
            }

            @Override
            public void onNothingSelected(final AdapterView<?> parent) {
                // do nothing
            }
        });
    }
}

File: app/src/main/java/com/meli/android/carddrawer/app/MainActivity.kt
package com.meli.android.carddrawer.app

import android.graphics.Color
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.util.Log
import android.view.View
import android.widget.ArrayAdapter
import android.widget.AutoCompleteTextView
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.SwitchCompat
import com.meli.android.carddrawer.app.SwitchFactoryModelSample.createModel
import com.meli.android.carddrawer.app.SwitchFactoryModelSample.createModelForAccountMoneyHybrid
import com.meli.android.carddrawer.app.model.CardComposite
import com.meli.android.carddrawer.app.model.CustomAccountMoneyConfiguration
import com.meli.android.carddrawer.app.model.HybridCreditConfiguration
import com.meli.android.carddrawer.app.model.MasterCardConfiguration
import com.meli.android.carddrawer.app.model.PixConfiguration
import com.meli.android.carddrawer.app.model.UrlTestConfiguration
import com.meli.android.carddrawer.app.model.VisaCardBlueConfiguration
import com.meli.android.carddrawer.app.model.VisaCardGrayConfiguration
import com.meli.android.carddrawer.app.model.VisaCardGreenConfiguration
import com.meli.android.carddrawer.app.model.VisaCardRedConfiguration
import com.meli.android.carddrawer.app.model.VisaCardYellowConfiguration
import com.meli.android.carddrawer.configuration.CardDrawerStyle
import com.meli.android.carddrawer.configuration.DefaultCardConfiguration
import com.meli.android.carddrawer.model.CardDrawerSource
import com.meli.android.carddrawer.model.CardDrawerView
import com.meli.android.carddrawer.model.Label
import com.meli.android.carddrawer.model.customview.CardDrawerSwitch

class MainActivity : AppCompatActivity() {

    private lateinit var switchCustomView: SwitchCompat
    private lateinit var switchTag: SwitchCompat
    private val card = CardComposite()
    private val cardViews = mutableListOf<CardDrawerViewOption>()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.card_drawer_app_activity_main)

        addCard("High", findViewById(R.id.card_header_container))
        addCard("Lowres", findViewById(R.id.card_header_lowres_container))
        addCard("Medium", findViewById(R.id.card_header_medium_container))
        addCard("MediumRes", findViewById(R.id.card_header_mediumres_container))
        addCard("Small", findViewById(R.id.card_header_small_container))

        initSizesDropDown()

        switchTag = findViewById(R.id.card_header_show_tag_switch)
        switchTag.setOnCheckedChangeListener { _, _ ->
            updateCardTagVisibility()
        }

        switchCustomView = findViewById(R.id.card_drawer_custom_view_switch)
        switchCustomView.setOnCheckedChangeListener { _, _ ->
            setUpCustomView(null)
        }

        findViewById<SwitchCompat>(R.id.card_header_switch_responsive).setOnCheckedChangeListener { _, isChecked ->
            val behaviour = if (isChecked) CardDrawerView.Behaviour.RESPONSIVE else CardDrawerView.Behaviour.REGULAR
            updateCardBehaviour(behaviour)
        }

        findViewById<SwitchCompat>(R.id.card_header_show_bottom_label_switch).setOnCheckedChangeListener { _, isChecked ->
            if (isChecked) {
                showBottomLabel()
            } else {
                hideBottomLabel()
            }
        }

        findViewById<SwitchCompat>(R.id.card_header_disabled_switch).setOnCheckedChangeListener { _, isChecked ->
            cardViews.forEach {
                it.view.isEnabled = !isChecked
            }
        }

        initCardConfigurationOptions()
        initCardNumber()
        initCardName()
        initExpirationDate()
        initSecurityCode()
    }

    private fun showBottomLabel() {
        val label = Label("mensaje destacado")
        cardViews.forEach {
            it.view.setBottomLabel(label)
            it.view.showBottomLabel()
        }
    }

    private fun hideBottomLabel() {
        cardViews.forEach {
            it.view.hideBottomLabel()
        }
    }

    private fun initSecurityCode() {
        findViewById<TextView>(R.id.security_code).addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {
                showSecurityCode()
            }

            override fun onTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {
                card.secCode = charSequence.toString()
            }

            override fun afterTextChanged(editable: Editable) {
                if (editable.toString().isEmpty()) {
                    showCard()
                }
            }
        })
    }

    private fun initExpirationDate() {
        findViewById<TextView>(R.id.expiration_date).addTextChangedListener(object : CustomTextWatcher() {
            override fun onTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {
                card.expiration = charSequence.toString()
            }
        })
    }

    private fun initCardName() {
        val cardName = findViewById<TextView>(R.id.card_name)
        cardName.addTextChangedListener(object : CustomTextWatcher() {
            override fun onTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {
                card.name = charSequence.toString()
            }
        })
        cardName.setOnClickListener {
            showCard()
        }
    }

    private fun initCardNumber() {
        val cardNumber = findViewById<TextView>(R.id.card_number)
        cardNumber.addTextChangedListener(object : CustomTextWatcher() {
            override fun onTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {
                card.number = charSequence.toString()
            }
        })
        cardNumber.setOnClickListener {
            showCard()
        }
    }

    private fun initCardConfigurationOptions() {
        val cardConfigurations = findViewById<AutoCompleteTextView>(R.id.configurations)
        val cardOptions = mutableListOf<CardConfigurationOption>()
        val tag = CardDrawerSource.Tag(
            "Novo",
            Color.parseColor("#1A4189E6"), Color.parseColor("#009EE3"), "regular"
        )
        val defaultOption = CardConfigurationOption("Default", DefaultCardConfiguration(this), tag)
        cardOptions.add(defaultOption)
        cardOptions.add(CardConfigurationOption("Visa blue", VisaCardBlueConfiguration(this), tag))
        cardOptions.add(CardConfigurationOption("Visa green", VisaCardGreenConfiguration(this), tag))
        cardOptions.add(CardConfigurationOption("Visa red", VisaCardRedConfiguration(this), tag))
        cardOptions.add(CardConfigurationOption("Visa gray", VisaCardGrayConfiguration(this), tag))
        cardOptions.add(CardConfigurationOption("Visa yellow", VisaCardYellowConfiguration(this), tag))
        cardOptions.add(CardConfigurationOption("Master", MasterCardConfiguration(this), tag))
        cardOptions.add(CardConfigurationOption("Url Test", UrlTestConfiguration(this), tag))
        cardOptions.add(CardConfigurationOption("Hybrid Account Money", CardDrawerStyle.ACCOUNT_MONEY_HYBRID, tag))
        cardOptions.add(CardConfigurationOption("Hybrid Credit", HybridCreditConfiguration(), tag))
        cardOptions.add(CardConfigurationOption("Default Account Money", CardDrawerStyle.ACCOUNT_MONEY_DEFAULT, tag))
        cardOptions.add(CardConfigurationOption("Custom account money", CustomAccountMoneyConfiguration(), tag))
        cardOptions.add(CardConfigurationOption("PIX", PixConfiguration(this)))
        val cardAdapter = ArrayAdapter(this, R.layout.card_drawer_list_item, cardOptions)
        cardConfigurations.setAdapter(cardAdapter)
        cardConfigurations.setOnItemClickListener { _, _, position, _ ->
            val selection = cardOptions[position]
            val source = selection.cardConfiguration
            val style = selection.cardStyle
            if (source != null) {
                updateCardSource(source)
            } else if (style != null) {
                updateCardStyle(style, selection.styledCardTag)
            }
            updateCardTagVisibility()
            setUpCustomView(style)
        }
        cardConfigurations.setText(defaultOption.toString(), false)
        updateCardSource(defaultOption.cardConfiguration!!)
    }

    private fun setUpCustomView(style: CardDrawerStyle?) {
        cardViews.forEach {
            if (switchCustomView.isChecked) {
                val model = if (style === CardDrawerStyle.ACCOUNT_MONEY_HYBRID) {
                    createModelForAccountMoneyHybrid()
                } else {
                    createModel()
                }

                val switchView = CardDrawerSwitch(this@MainActivity)
                switchView.setSwitchListener(object : CardDrawerSwitch.OnSwitchListener {
                    override fun onChange(id: String) {
                        Log.i(
                            "CARD_DRAWER",
                            "ID: $id"
                        )
                    }
                })
                switchView.setSwitchModel(model)
                switchView.setConfiguration(it.view.buildCustomViewConfiguration())
                it.view.setCustomView(switchView)
            } else {
                it.view.setCustomView(null)
            }
        }
    }

    private fun initSizesDropDown() {
        findViewById<AutoCompleteTextView>(R.id.sizes).also {
            it.setAdapter(ArrayAdapter(this, R.layout.card_drawer_list_item, cardViews))
            it.setText(cardViews.first().name, false)
            it.setOnItemClickListener { _, _, position, _ ->
                updateCardVisibility(cardViews[position].view)
            }
        }
    }

    private fun updateCardTagVisibility() {
        val visibility = if (switchTag.isChecked) View.VISIBLE else View.GONE
        cardViews.forEach {
            val regularTag =
                it.view.findViewById<View>(R.id.card_header_front).findViewById<View>(R.id.card_tag_container)
            val genericTag =
                it.view.findViewById<View>(R.id.card_drawer_generic_front).findViewById<View>(R.id.card_tag_container)
            regularTag.visibility = visibility
            genericTag.visibility = visibility
        }
    }

    private fun updateCardVisibility(visibleCard: CardDrawerView) {
        cardViews.forEach {
            it.view.visibility = if (it.view == visibleCard) View.VISIBLE else View.GONE
        }
    }

    private fun showSecurityCode() {
        cardViews.forEach {
            it.view.showSecurityCode()
        }
    }

    private fun showCard() {
        cardViews.forEach {
            it.view.show()
        }
    }

    private fun updateCardSource(source: CardDrawerSource) {
        cardViews.forEach {
            it.view.show(source)
        }
    }

    private fun updateCardBehaviour(behaviour: Int) {
        cardViews.forEach {
            it.view.setBehaviour(behaviour)
        }
    }

    private fun updateCardStyle(style: CardDrawerStyle, tag: CardDrawerSource.Tag?) {
        cardViews.forEach {
            it.view.setStyle(style, tag)
        }
    }

    private fun addCard(key:String, cardDrawerView: CardDrawerView) {
        cardViews.add(CardDrawerViewOption(key, cardDrawerView))
        card.addCard(cardDrawerView.card)
    }

    abstract class CustomTextWatcher : TextWatcher {
        override fun beforeTextChanged(charSequence: CharSequence, start: Int, count: Int, after: Int) = Unit
        override fun afterTextChanged(editable: Editable) = Unit
    }
}
